controllers/global_search_setup.go

```diff
@@ -0,0 +1,554 @@
+		Group:    "multicluster.openshift.io",
+		Version:  "v1",
+		Resource: "multiclusterengines",
+	}
+	managedClusterResourceGvr = schema.GroupVersionResource{
+		Group:    "cluster.open-cluster-management.io",
+		Version:  "v1",
+		Resource: "managedclusters",
+	}
+	managedServiceAccountGvr = schema.GroupVersionResource{
+		Group:    "authentication.open-cluster-management.io",
+		Version:  "v1beta1",
+		Resource: "managedserviceaccounts",
+	}
+	manifestWorkGvr = schema.GroupVersionResource{
+		Group:    "work.open-cluster-management.io",
+		Version:  "v1",
+		Resource: "manifestworks",
+	}
+)
+
+// Reconcile Global Search.
+//  1. Check the global-search-preview annotation.
+//  2. Validate pre-requisites.
+//     a. MulticlusterGlobalHub operator is installed in the cluster.
+//     b. The ManagedServiceAccount add-on is enabled in the MultiClusterEngine CR.
+//     c. The ClusterProxy addon is enabled in the MultiClusterEngine CR.
+//  3. Enable global search feature in the console.
+//  4. Enable federated search feature in the search-api deployment.
+//  5. Create a ManagedServiceAccount recource for each Managed Hub.
+//  6. Create a ManifestWork resource for each Managed Hub.
+func (r *SearchReconciler) reconcileGlobalSearch(ctx context.Context,
+	instance *searchv1alpha1.Search) (*reconcile.Result, error) {
+
+	if instance.ObjectMeta.Annotations["search.open-cluster-management.io/global-search-peview"] == "true" ||
+		instance.ObjectMeta.Annotations["global-search-preview"] == "true" {
+
+		log.Info("The global-search-preview annotation is present. Setting up global search...")
+
+		// Validate global search pre-requisites.
+		err := r.validateGlobalSearchPrerequisites(ctx)
+		if err != nil {
+			log.Error(err, "Failed to verify global search pre-requisites.")
+			return &reconcile.Result{}, err
+		}
+
+		// Enable global search.
+		err = r.enableGlobalSearch(ctx, instance)
+		if err != nil {
+			log.Info("Failed to enable global search. Updating CR status conditions.", "error", err.Error())
+
+			updateErr := r.updateGlobalSearchStatus(ctx, instance, metav1.Condition{
+				Type:               "GlobalSearchReady",
+				Status:             metav1.ConditionFalse,
+				Reason:             "GlobalSearchSetupFailed",
+				Message:            "Failed to enable global search. " + err.Error(),
+				LastTransitionTime: metav1.Now(),
+			})
+			if updateErr != nil {
+				log.Error(updateErr, "Failed to update Global Search status condition on Search CR instance.")
+			}
+
+		} else {
+			updateErr := r.updateGlobalSearchStatus(ctx, instance, metav1.Condition{
+				Type:               "GlobalSearchReady",
+				Status:             metav1.ConditionTrue,
+				Reason:             "None",
+				Message:            "None",
+				LastTransitionTime: metav1.Now(),
+			})
+			if updateErr != nil {
+				log.Error(updateErr, "Failed to update the global search status condition on Search CR instance.")
```

Comment:
```suggestion
				log.Error(updateErr, "Failed to update the Global Search status condition on Search CR instance.")
```
To be same as above.

Reply:
done